### PR TITLE
PatchBufferOp issues when lowering load ptr(7), ptr(7)

### PR DESCRIPTION
**Pull request by ruiminzhao**
##Issue Description
There are issue when lowering on load instruction which both result type and pointer operand type are fat buffer pointer.

An example IR:
```
....
  %6 = call ptr addrspace(7) @lgc.buffer.desc.to.ptr(<4 x i32> %5)

  %10 = load ptr addrspace(7), ptr addrspace(7) %6, align 32
  %11 = load i8, ptr addrspace(7) %10, align 1
  store i8 %11, ptr addrspace(7) %9, align 1
```

Currently, as %10 and %11 which operand address space are 7,  so they are all be added into postVisitLoad list.
The 1st issue is when call replaceLoadStore for %10, it will be lowered into i160 not <4xi32, i32> finally.
Then call replaceLoadStore for %11, it will be crashed when call m_typeLowering.getValue(pointerOperand).

The 2nd issue is TypeLowering.visitLoad will handle on %10 firstly, then BufferOp.postVisitLoad will handle on %10 again 
as the pointer operand address space is 7(fat_buffer_pointer), then %10 will be added into m_instructionsToErase again, then it will be crashed when erase this load instruction.

##Issue Solution:
Now I have moved the instructions formatted "load ptr7, ptr7" from typeLowering.visitLoad into PatchBufferOp.postVisitLoad, but I'm not sure this will be correct for other parts such as CPS, which doesn't have postVisitLoad function to handle operand = ptr<7> cases.

Any better solution will be appreciated.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 284edc5b555708d7687cf9e15aa9d6bd660c65d4
<details><summary>CTS tests (Failed: 0/138443)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35211/69228 (50.9%)</li>
<li>Failed: 0/69228 (0.0%)</li>
<li>Not Supported: 34017/69228 (49.1%)</li><li>Warnings: 0/69228 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35241/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33974/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit ccac591f6d9d568fe59bf8b9d0d5162e2a5f8214
<details><summary>CTS tests (Failed: 0/138378)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35162/69163 (50.8%)</li>
<li>Failed: 0/69163 (0.0%)</li>
<li>Not Supported: 34001/69163 (49.2%)</li><li>Warnings: 0/69163 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35241/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33974/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
I'll take a look, but how is it possible for you to end up with such code in the first place? What's the corresponding SPIR-V code that would cause you to load or store a buffer pointer in a buffer?

**Review by nhaehnle — COMMENTED** [review]
Okay, we need to understand better whether and why this is happening to support this. A couple of strategies:

One idea is to scan the entire code multiple times, with the BufferOpLowering visit of `load` and `store` happening during the second scan. But that is bad because the second scan is usually not required and it costs compile time.

Another idea is to teach the visitor infrastructure about potentially having to re-visit newly inserted instructions. Basically, TypeLowering::visitLoad would return some indication that the code which has just been created by it has to be visited again. This would mean that visitor functions can return some `llvm_dialects::VisitResult` value.

Cc @JanRehders-AMD

File: compilerutils/lib/TypeLowering.cpp [hunk 1]
```diff
@@ -560,6 +560,12 @@ void TypeLowering::visitInsert(llvm::InsertValueInst &insert) {
 //
 // @param load : the instruction
 void TypeLowering::visitLoad(LoadInst &load) {
+  // For the pointerOperand which address space is ADDR_SPACE_BUFFER_FAT_POINTER, it will be lowered
+  // on postVisitLoad. But this maybe have impacts on other pass like CPS which doesn't have postVisitLoad
+  if (load.getPointerAddressSpace() == 7) {
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (ruiling) [reply compilerutils/lib/TypeLowering.cpp:567]: While the `TypeLowering::visitLoad` is useful in general case. But it does not quite well with PatchBufferOp for such tricky case. Maybe we can just derive `BufferOpLowering` from `TypeLowering`, and only register `BufferOpLowering::visitLoad` without `TypeLowering::visitLoad`. Can we? The load of fat pointer from memory pointed by fat pointer would be handled solely by `BufferOpLowering::postVisitLoad` which is just like the idea in the change.

**Comment by ruiling** [comment]
> Another idea is to teach the visitor infrastructure about potentially having to re-visit newly inserted instructions. Basically, TypeLowering::visitLoad would return some indication that the code which has just been created by it has to be visited again. This would mean that visitor functions can return some `llvm_dialects::VisitResult` value.

Besides revisiting newly inserted instructions, we also need a way to notify other visitors stop visiting the old instruction which was just being lowered. For this case, the `TypeLowering::visitLoad` needs a way to notify `BufferOpLowering`: the load instruction has been lowered and marked for deletion, please don't visit/lower it again. I am not sure whether this can be achieved somehow and whether this is doable.

File: compilerutils/lib/TypeLowering.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ruiling) [reply compilerutils/lib/TypeLowering.cpp:567]: A more proper way might be to explicitly abort on such kind of complex case in `TypeLowering::visitLoad` and leave it to other visitor to handle it.
```diff
--- a/compilerutils/lib/TypeLowering.cpp
+++ b/compilerutils/lib/TypeLowering.cpp
@@ -488,6 +488,13 @@ void TypeLowering::visitLoad(LoadInst &load) {
   if (types.size() == 1 && types[0] == load.getType())
     return;

+  auto *pointerType = load.getPointerOperand()->getType();
+  auto pointerTypes = convertType(pointerType);
+  // This is somewhat complex that we are lowering both the pointer and the loaded value, skip this to let other visitor
+  // handle it more properly.
+  if (pointerTypes[0] != pointerType)
+    return;
+
   m_builder.SetInsertPoint(&load);

   Type *loadType;
```

**Comment by ruiminzhao** [comment]
> but how is it possible for you to end up with such code in the first place

This case has been generated by VKI_KHR_SHADER_UNTYPED_POINTERS CTS case:
https://gerrit.khronos.org/c/vk-gl-cts/+/10647. As this extension hasn't been published, I just copy the LLVM-IR here.
The original SPIRV code is:
```
OpCapability UntypedPointersKHR
OpCapability UntypedPointersStorageBufferKHR


OpCapability Int8


OpCapability VariablePointersStorageBuffer
OpCapability VariablePointers
OpCapability VulkanMemoryModel
OpCapability VulkanMemoryModelDeviceScopeKHR
OpExtension &quot;SPV_KHR_vulkan_memory_model&quot;

OpExtension &quot;SPV_KHR_storage_buffer_storage_class&quot;
OpExtension &quot;SPV_KHR_untyped_pointers&quot;
OpExtension &quot;SPV_KHR_variable_pointers&quot;

OpMemoryModel Logical Vulkan
OpEntryPoint GLCompute %main &quot;main&quot; %id  
OpExecutionMode %main LocalSize 1 1 1
OpDecorate %id BuiltIn GlobalInvocationId
OpMemberDecorate %input_buffer_0           0             Offset 0
OpDecorate       %input_buffer_0           Block
OpDecorate       %input_data_0_untyped_var DescriptorSet 0
OpDecorate       %input_data_0_untyped_var Binding       0
OpMemberDecorate %input_buffer_1           0             Offset 0
OpDecorate       %input_buffer_1           Block
OpDecorate       %input_data_1_untyped_var DescriptorSet 0
OpDecorate       %input_data_1_untyped_var Binding       1
OpMemberDecorate %output_buffer            0             Offset 0
OpDecorate       %output_buffer            Block
OpDecorate       %output_data_var          DescriptorSet 0
OpDecorate       %output_data_var          Binding       2
%uint32     = OpTypeInt  32      0
%c_uint32_1 = OpConstant %uint32 1
%void         = OpTypeVoid
%bool         = OpTypeBool
%uint8  = OpTypeInt    8 0
%vec3_uint32  = OpTypeVector %uint32 3
%void_func   = OpTypeFunction %void
%c_uint32_0  = OpConstant %uint32 0
%c_uint32_16 = OpConstant %uint32 16
%c_bool_true = OpConstantTrue %bool
%uint32_input_ptr                 = OpTypePointer Input %uint32
%vec3_uint32_input_ptr            = OpTypePointer Input %vec3_uint32
%input_buffer_0                          = OpTypeStruct %uint8
%input_buffer_1                          = OpTypeStruct %uint8
%output_buffer                           = OpTypeStruct %uint8
%uint8_storage_buffer_ptr          = OpTypePointer           StorageBuffer                     %uint8
%output_buffer_storage_buffer_ptr        = OpTypePointer           StorageBuffer                     %output_buffer
%storage_buffer_untyped_ptr              = OpTypeUntypedPointerKHR StorageBuffer
%storage_buffer_untyped_ptr_private_ptr  = OpTypePointer           Private                           %storage_buffer_untyped_ptr
%input_data_0_untyped_var                = OpUntypedVariableKHR    %storage_buffer_untyped_ptr       %input_buffer_0 StorageBuffer
%input_data_1_untyped_var                = OpUntypedVariableKHR    %storage_buffer_untyped_ptr       %input_buffer_1 StorageBuffer
%output_data_var                         = OpVariable              %output_buffer_storage_buffer_ptr                 StorageBuffer
%output_copy_private_var                 = OpVariable              %storage_buffer_untyped_ptr_private_ptr           Private
%id              = OpVariable           %vec3_uint32_input_ptr                            Input
%main       = OpFunction %void None %void_func
%label_main = OpLabel
%input_loc_first          = OpUntypedAccessChainKHR    %storage_buffer_untyped_ptr              %input_buffer_0           %input_data_0_untyped_var %c_uint32_0
%input_loc_second         = OpUntypedAccessChainKHR    %storage_buffer_untyped_ptr              %input_buffer_1           %input_data_1_untyped_var %c_uint32_0
%output_loc               = OpAccessChain              %uint8_storage_buffer_ptr                                    %output_data_var          %c_uint32_0
%selected_ptr             = OpSelect                   %storage_buffer_untyped_ptr              %c_bool_true              %input_loc_first          %input_loc_second
                            OpStore                    %output_copy_private_var                 %selected_ptr
%output_copy_loc_unty_ptr = OpLoad                     %storage_buffer_untyped_ptr              %output_copy_private_var
%output_copy_loc_ptr      = OpLoad                     %uint8_storage_buffer_ptr          %output_copy_loc_unty_ptr
%output_copy_loc          = OpLoad                     %uint8                             %output_copy_loc_ptr
                            OpStore                    %output_loc                              %output_copy_loc
                OpReturn
                OpFunctionEnd
```

**Comment by ruiminzhao** [comment]
Close this PR and new PR submitted which only added the assertion in https://github.com/GPUOpen-Drivers/llpc/pull/2881.

Any case meet this pattern: "load ptr(7), ptr(7)", I will reopen this PR for fix. 

Thanks.

**Pull request closed without merging**